### PR TITLE
refactoring dataloader into registries.

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
Training run for text dataloader before and after refactoring overlap:
1. Loss :Screenshot 2024-05-02 at 1 08 06 PM
2. MFU : Screenshot 2024-05-02 at 1 08 27 PM
3. Memory : Screenshot 2024-05-02 at 1 08 40 PM

Training run for finetuning dataloader with streaming dataset before and after refactoring overlap:
1. Loss :
Screenshot 2024-05-02 at 1 10 46 PM
2. MFU : 
Screenshot 2024-05-02 at 1 11 23 PM
3. Memory : 
Screenshot 2024-05-02 at 1 10 30 PM

Training run for finetuning dataloader with huggingface dataset before and after refactoring overlap:
1. Loss :
Screenshot 2024-05-02 at 1 12 16 PM

2. MFU : 
Screenshot 2024-05-02 at 1 12 26 PM

3. Memory : 

Screenshot 2024-05-02 at 1 12 47 PM

### CODE REVIEW

File: llmfoundry/data/text_data.py [hunk 1]
```diff
@@ -287,73 +284,56 @@ def build_streams(dataset_cfg: DictConfig):
...
-    device_batch_size: int,
+    device_batch_size: Union[int, float],
 ) -> DataSpec:
     assert cfg.name == 'text', f'Tried to build text dataloader with cfg.name={cfg.name}'
 
     # get kwargs
-    mlm_probability = cfg.dataset.pop('mlm_probability', None)
-    eos_token_id = cfg.dataset.pop('eos_token_id', None)
-    bos_token_id = cfg.dataset.pop('bos_token_id', None)
+    cfg.dataset['replication'], ds_batch_size = construct_from_registry(
```
Comment (dakinggg) [inline llmfoundry/data/text_data.py:292]: minor nit, spell out `dataset_batch_size`

File: llmfoundry/registry.py [hunk 2]
```diff
@@ -137,6 +138,61 @@
...
+
+_collators_description = (
+    """Returns the data collator.
+    Args:
+        cfg (DictConfig): The dataloader config.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer
+        ds_batch_size (Union[int, float]): The dataset device batch size.
+    Returns:
+        collate_fn  (Any): The collate function.
+        dataloader_batch_size (Optional[int]): The dataloader batch size, used for packing. Only finetuning collator returns this."""
```
Comment (dakinggg) [inline llmfoundry/registry.py:169]: they should all return this. if it doesn't need to change it, it should just be a no op on the input, right?

File: llmfoundry/registry.py [hunk 3]
```diff
@@ -137,6 +138,61 @@
...
+    description=_collators_description,
+)
+
+_data_specs_description = (
+    """Returns the get_data_spec function.
+    Args:
+        dl (Union[Iterable, TorchDataloader): The dataloader.
+        dataset_cfg (DictConfig): The dataset config.
+    Returns:
+        dataspec  (Any): The dataspec."""
```
Comment (dakinggg) [inline llmfoundry/registry.py:186]: datspec should be typed as dataspec, right?

**Review by dakinggg — APPROVED** [review]
Thank you!

File: llmfoundry/registry.py (hunk 2 above)
Comment (ShashankMosaicML) [reply llmfoundry/registry.py:169]: Right. Fixed.
